File: dcf-app/frontend/src/api/dcf.ts
import { apiFetch } from "./client";
import type {
  DcfAssumptions,
  DcfCalculationResponse,
  ReverseDcfRequest,
  ReverseDcfResponse,
  SavedModelPayload,
  SavedModelSummary,
  SavedValuationModel,
  ScenarioCompareResponse,
  SensitivityMatrixResponse,
  SensitivityRequest,
  TargetReturnRequest,
  TargetReturnResponse,
} from "../types/dcf";

export function calculateDcf(assumptions: DcfAssumptions) {
  return apiFetch<DcfCalculationResponse>("/dcf/calculate", {
    method: "POST",
    body: JSON.stringify(assumptions),
  });
}

export function calculateSensitivity(payload: SensitivityRequest) {
  return apiFetch<SensitivityMatrixResponse>("/dcf/sensitivity", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function compareScenarios(scenarios: DcfAssumptions[]) {
  return apiFetch<ScenarioCompareResponse>("/dcf/scenario/compare", {
    method: "POST",
    body: JSON.stringify({ scenarios }),
  });
}

export function reverseDcf(payload: ReverseDcfRequest) {
  return apiFetch<ReverseDcfResponse>("/dcf/reverse", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function targetReturn(payload: TargetReturnRequest) {
  return apiFetch<TargetReturnResponse>("/dcf/target-return", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function listModels() {
  return apiFetch<SavedModelSummary[]>("/models");
}

export function getModel(id: string) {
  return apiFetch<SavedValuationModel>(`/models/${id}`);
}

export function createModel(payload: SavedModelPayload) {
  return apiFetch<SavedValuationModel>("/models", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function updateModel(id: string, payload: SavedModelPayload) {
  return apiFetch<SavedValuationModel>(`/models/${id}`, {
    method: "PUT",
    body: JSON.stringify(payload),
  });
}

export function deleteModel(id: string) {
  return apiFetch<void>(`/models/${id}`, {
    method: "DELETE",
  });
}

export function duplicateModel(id: string) {
  return apiFetch<SavedValuationModel>(`/models/${id}/duplicate`, {
    method: "POST",
  });
}

File: dcf-app/backend/app/api/routes/dcf.py
from fastapi import APIRouter, Depends

from app.api.dependencies import get_dcf_service
from app.schemas.dcf import (
    DcfAssumptions,
    DcfCalculationResponse,
    ReverseDcfRequest,
    ReverseDcfResponse,
    ScenarioCompareRequest,
    ScenarioCompareResponse,
    SensitivityMatrixResponse,
    SensitivityRequest,
    TargetReturnRequest,
    TargetReturnResponse,
)
from app.services.dcf_service import DcfService


router = APIRouter()


@router.post("/calculate", response_model=DcfCalculationResponse)
def calculate_dcf(
    assumptions: DcfAssumptions,
    service: DcfService = Depends(get_dcf_service),
) -> DcfCalculationResponse:
    return service.calculate(assumptions)


@router.post("/sensitivity", response_model=SensitivityMatrixResponse)
def calculate_sensitivity(
    request: SensitivityRequest,
    service: DcfService = Depends(get_dcf_service),
) -> SensitivityMatrixResponse:
    return service.build_sensitivity(request)


@router.post("/scenario/compare", response_model=ScenarioCompareResponse)
def compare_scenarios(
    request: ScenarioCompareRequest,
    service: DcfService = Depends(get_dcf_service),
) -> ScenarioCompareResponse:
    return service.compare_scenarios(request)


@router.post("/reverse", response_model=ReverseDcfResponse)
def reverse_dcf(
    request: ReverseDcfRequest,
    service: DcfService = Depends(get_dcf_service),
) -> ReverseDcfResponse:
    return service.reverse_dcf(request)


@router.post("/target-return", response_model=TargetReturnResponse)
def target_return(
    request: TargetReturnRequest,
    service: DcfService = Depends(get_dcf_service),
) -> TargetReturnResponse:
    return service.target_return(request)

